File: src/dataCodec.cpp
#include "dataCodec.h"

template <> TYPE dataCodecBase::_identifyType<bool>() {return TYPE::BOOL;};
template <> TYPE dataCodecBase::_identifyType<uint8_t>() {return TYPE::UINT;};
template <> TYPE dataCodecBase::_identifyType<uint16_t>() {return TYPE::UINT;};
template <> TYPE dataCodecBase::_identifyType<uint32_t>() {return TYPE::UINT;};
template <> TYPE dataCodecBase::_identifyType<uint64_t>() {return TYPE::UINT;};
template <> TYPE dataCodecBase::_identifyType<int8_t>() {return TYPE::INT;};
template <> TYPE dataCodecBase::_identifyType<int16_t>() {return TYPE::INT;};
template <> TYPE dataCodecBase::_identifyType<int32_t>() {return TYPE::INT;};
template <> TYPE dataCodecBase::_identifyType<int64_t>() {return TYPE::INT;};
template <> TYPE dataCodecBase::_identifyType<float>() {return TYPE::FLOAT;};
template <> TYPE dataCodecBase::_identifyType<double>() {return TYPE::FLOAT;};

template <typename T>
uint8_t dataCodecBase::_decideSize(TYPE type_, uint8_t size_) {
  if (type_ == TYPE::BOOL) return 1;
  if (type_ == TYPE::FLOAT) return 32;
  return size_;
}
template uint8_t dataCodecBase::_decideSize<bool>(TYPE type_, uint8_t size_);
template uint8_t dataCodecBase::_decideSize<uint8_t>(TYPE type_, uint8_t size_);
template uint8_t dataCodecBase::_decideSize<uint16_t>(TYPE type_, uint8_t size_);
template uint8_t dataCodecBase::_decideSize<uint32_t>(TYPE type_, uint8_t size_);
template uint8_t dataCodecBase::_decideSize<uint64_t>(TYPE type_, uint8_t size_);
template uint8_t dataCodecBase::_decideSize<int8_t>(TYPE type_, uint8_t size_);
template uint8_t dataCodecBase::_decideSize<int16_t>(TYPE type_, uint8_t size_);
template uint8_t dataCodecBase::_decideSize<int32_t>(TYPE type_, uint8_t size_);
template uint8_t dataCodecBase::_decideSize<int64_t>(TYPE type_, uint8_t size_);
template uint8_t dataCodecBase::_decideSize<float>(TYPE type_, uint8_t size_);
template uint8_t dataCodecBase::_decideSize<double>(TYPE type_, uint8_t size_);

#ifndef DATA_DECODER_ONLY
dataEncoder::dataEncoder(uint8_t id_){
  _dataSet.isEditable = true;
  _dataSet.id = id_;
  _dataPacket.id = id_;
};

template <typename T>
ERROR dataEncoder::append(uint8_t ord_, T *dataPtr_, uint8_t size_) {
  if (!_dataSet.isEditable) return ERROR::UNEDITABLE;
  if (MAX_DATA_NUM <= ord_) return ERROR::INVALID_PARAM;
  if (_dataSet.info[ord_].isActive) return ERROR::INVALID_PARAM;
  if (dataPtr_ == nullptr) return ERROR::INVALID_PARAM;
  if ((size_ == 0) || (32 < size_)) return ERROR::INVALID_PARAM;
  _dataSet.info[ord_].isActive = true;
  _dataSet.info[ord_].ptr = static_cast<void*>(dataPtr_);
  _dataSet.info[ord_].type = _identifyType<T>();
  _dataSet.info[ord_].size.raw = sizeof(T) * 8;
  _dataSet.info[ord_].size.encoded = _decideSize<T>(_dataSet.info[ord_].type, size_);
  return ERROR::OK;
}
template ERROR dataEncoder::append<bool>(uint8_t ord_, bool *dataPtr_, uint8_t size_);
template ERROR dataEncoder::append<uint8_t>(uint8_t ord_, uint8_t *dataPtr_, uint8_t size_);
template ERROR dataEncoder::append<uint16_t>(uint8_t ord_, uint16_t *dataPtr_, uint8_t size_);
template ERROR dataEncoder::append<uint32_t>(uint8_t ord_, uint32_t *dataPtr_, uint8_t size_);
template ERROR dataEncoder::append<uint64_t>(uint8_t ord_, uint64_t *dataPtr_, uint8_t size_);
template ERROR dataEncoder::append<int8_t>(uint8_t ord_, int8_t *dataPtr_, uint8_t size_);
template ERROR dataEncoder::append<int16_t>(uint8_t ord_, int16_t *dataPtr_, uint8_t size_);
template ERROR dataEncoder::append<int32_t>(uint8_t ord_, int32_t *dataPtr_, uint8_t size_);
template ERROR dataEncoder::append<int64_t>(uint8_t ord_, int64_t *dataPtr_, uint8_t size_);
template ERROR dataEncoder::append<float>(uint8_t ord_, float *dataPtr_, uint8_t size_);
template ERROR dataEncoder::append<double>(uint8_t ord_, double *dataPtr_, uint8_t size_);

ERROR dataEncoder::set() {
  _dataSet.bitLength = 0;
  for (uint8_t i = 0; i < MAX_DATA_NUM; i++) if (_dataSet.info[i].isActive) _dataSet.bitLength += _dataSet.info[i].size.encoded;
  if (_dataSet.bitLength > DATA_LENGTH_LIMIT) return ERROR::OVERFLOW;
  _dataSet.isEditable = false;
  return ERROR::OK;
}

ERROR dataEncoder::encode() {
  for (uint8_t i = 0; i < MAX_BYTE; i++) _dataPacket.data[i] = 0;
  _generateBinary();
  _dataPacket.data[0] = (0b10000000 | _dataSet.id);
  uint16_t idx = 0;
  for (uint16_t i = 0; i < _dataSet.bitLength; i++) {
    _dataPacket.data[idx + 1] |= (_binary[i] << (6 - (i % 7)));
    if ((i % 7) == 6) idx++;
  }
  if ((_dataSet.bitLength % 7) != 0) idx++;
  if ((idx / 64) > 1) return ERROR::OVERFLOW;
  uint16_t checkSum = 0;
  for (uint8_t i = 1; i < (idx + 1); i++) checkSum += _dataPacket.data[i];
  if (checkSum > 0x3FFF) return ERROR::OVERFLOW;
  _dataPacket.data[idx + 1] = ((checkSum >> 7) & 0x007F);
  _dataPacket.data[idx + 2] = (checkSum & 0x007F);
  _dataPacket.data[idx + 3] = (0b11000000 | ((idx % 64) & 0x003F));
  _dataPacket.length = idx + 4;
  return ERROR::OK;
}

dataPacket_t dataEncoder::getPacket() {return _dataPacket;}

template <typename T>
uint32_t dataEncoder::_getBitsData(uint8_t ord_) {
  T value;
  if (_dataSet.info[ord_].type == TYPE::BOOL) value = (T)*static_cast<bool*>(_dataSet.info[ord_].ptr);
  if ((_dataSet.info[ord_].type == TYPE::UINT) && (_dataSet.info[ord_].size.raw == 8)) value = (T)*static_cast<uint8_t*>(_dataSet.info[ord_].ptr);
  if ((_dataSet.info[ord_].type == TYPE::UINT) && (_dataSet.info[ord_].size.raw == 16)) value = (T)*static_cast<uint16_t*>(_dataSet.info[ord_].ptr);
  if ((_dataSet.info[ord_].type == TYPE::UINT) && (_dataSet.info[ord_].size.raw == 32)) value = (T)*static_cast<uint32_t*>(_dataSet.info[ord_].ptr);
  if ((_dataSet.info[ord_].type == TYPE::UINT) && (_dataSet.info[ord_].size.raw == 64)) value = (T)*static_cast<uint64_t*>(_dataSet.info[ord_].ptr);
  if ((_dataSet.info[ord_].type == TYPE::INT) && (_dataSet.info[ord_].size.raw == 8)) value = (T)*static_cast<int8_t*>(_dataSet.info[ord_].ptr);
  if ((_dataSet.info[ord_].type == TYPE::INT) && (_dataSet.info[ord_].size.raw == 16)) value = (T)*static_cast<int16_t*>(_dataSet.info[ord_].ptr);
  if ((_dataSet.info[ord_].type == TYPE::INT) && (_dataSet.info[ord_].size.raw == 32)) value = (T)*static_cast<int32_t*>(_dataSet.info[ord_].ptr);
  if ((_dataSet.info[ord_].type == TYPE::INT) && (_dataSet.info[ord_].size.raw == 64)) value = (T)*static_cast<int64_t*>(_dataSet.info[ord_].ptr);
  if ((_dataSet.info[ord_].type == TYPE::FLOAT) && (_dataSet.info[ord_].size.raw == 32)) value = (T)*static_cast<float*>(_dataSet.info[ord_].ptr);
  if ((_dataSet.info[ord_].type == TYPE::FLOAT) && (_dataSet.info[ord_].size.raw == 64)) value = (T)*static_cast<double*>(_dataSet.info[ord_].ptr);
  uint32_t bits = 0;
  memcpy(&bits, &value, sizeof(uint32_t));
  return bits;
}
template uint32_t dataEncoder::_getBitsData<uint32_t>(uint8_t ord_);
template uint32_t dataEncoder::_getBitsData<int32_t>(uint8_t ord_);
template uint32_t dataEncoder::_getBitsData<float>(uint8_t ord_);

uint32_t dataEncoder::_getEncodedData(uint8_t ord_) {
  uint32_t encoded;
  uint32_t mask = 0;
  if (_dataSet.info[ord_].type == TYPE::INT) {
    for (uint8_t i = 0; i < (_dataSet.info[ord_].size.encoded - 1); i++) mask = ((mask << 1) | 0x00000001);
    bool signBit = ((_dataSet.info[ord_].data.bits >> 31) & 0x00000001);
    encoded = ((_dataSet.info[ord_].data.bits & mask) | (signBit << (_dataSet.info[ord_].size.encoded - 1)));
  }
  else {
    for (uint8_t i = 0; i < _dataSet.info[ord_].size.encoded; i++) mask = ((mask << 1) | 1);
    encoded = (_dataSet.info[ord_].data.bits & mask);
  }
  return encoded;
}

void dataEncoder::_generateBinary() {
  uint16_t idx = 0;
  for (uint8_t i = 0; i < MAX_DATA_NUM; i++) {
    if (_dataSet.info[i].isActive) {
      if (_dataSet.info[i].type == TYPE::INT) {
        _dataSet.info[i].data.bits = _getBitsData<int32_t>(i);
        _dataSet.info[i].data.encoded = _getEncodedData(i);
      }
      else if (_dataSet.info[i].type == TYPE::FLOAT) {
        _dataSet.info[i].data.bits = _getBitsData<float>(i);
        _dataSet.info[i].data.encoded = _getEncodedData(i);
      }
      else {
        _dataSet.info[i].data.bits = _getBitsData<uint32_t>(i);
        _dataSet.info[i].data.encoded = _getEncodedData(i);
      }
      for (uint8_t j = 0; j < _dataSet.info[i].size.encoded; j++) {
        _binary[idx] = (_dataSet.info[i].data.encoded >> (_dataSet.info[i].size.encoded - j - 1) & 0x00000001);
        idx++;
      }
    }
  }
}
#endif

#ifndef DATA_ENCODER_ONLY
dataDecoder::dataDecoder() {
  for (uint8_t i = 0; i < MAX_PORT_NUM; i++) {
    _dataSet[i].isEditable = true;
    _dataSet[i].id = i;
  }
  _bufferIndex = 0;
  _currentId = 255;
  for (uint16_t i = 0; i < BUFFER_SIZE; i++) _buffer[i] = 0;
}

template <typename T>
ERROR dataDecoder::append(uint8_t id_, uint8_t ord_, T *dataPtr_, uint8_t size_) {
  if (!_dataSet[id_].isEditable) return ERROR::UNEDITABLE;
  if (MAX_DATA_NUM <= ord_) return ERROR::INVALID_PARAM;
  if (_dataSet[id_].info[ord_].isActive) return ERROR::INVALID_PARAM;
  if (dataPtr_ == nullptr) return ERROR::INVALID_PARAM;
  if ((size_ == 0) || (32 < size_)) return ERROR::INVALID_PARAM;
  _dataSet[id_].info[ord_].isActive = true;
  _dataSet[id_].info[ord_].ptr = static_cast<void*>(dataPtr_);
  _dataSet[id_].info[ord_].type = _identifyType<T>();
  _dataSet[id_].info[ord_].size.raw = sizeof(T) * 8;
  _dataSet[id_].info[ord_].size.encoded = _decideSize<T>(_dataSet[id_].info[ord_].type, size_);
  return ERROR::OK;
}
template ERROR dataDecoder::append<bool>(uint8_t id_, uint8_t ord_, bool *dataPtr_, uint8_t size_);
template ERROR dataDecoder::append<uint8_t>(uint8_t id_, uint8_t ord_, uint8_t *dataPtr_, uint8_t size_);
template ERROR dataDecoder::append<uint16_t>(uint8_t id_, uint8_t ord_, uint16_t *dataPtr_, uint8_t size_);
template ERROR dataDecoder::append<uint32_t>(uint8_t id_, uint32_t ord_, uint32_t *dataPtr_, uint8_t size_);

File: python/data_codec/data_codec.py
# data_codec.py
# C++ dataCodec 準拠版
#
# - dataEncoder: 1インスタンス = 1パケットID（id）
# - dataDecoder: 1インスタンス = 1固定パケットID
# - プロトコル:
#   header: 10iiiiii (i = id)
#   payload: 7bit pack (MSB→LSB, data[1] の bit6 から)
#   checksum: payload バイトの和の下位14bit → hi7 / lo7
#   tail: 11llllll (l = payload_length_bytes % 64)

from enum import Enum
from typing import Any, List, Optional
import struct

MAX_DATA_NUM = 10
MAX_BIT = MAX_DATA_NUM * 32
MAX_BYTE = (MAX_BIT + 6) // 7 + 4
BUFFER_SIZE = 4096
# ★DATA_LENGTH_LIMIT は「payload の最大バイト数」
DATA_LENGTH_LIMIT = 127

class TYPE(Enum):
    BOOL = 0
    UINT = 1
    INT = 2
    FLOAT = 3   # float / double（エンコード時は常に32bit）

class ERROR(Enum):
    OK = 0
    UNEDITABLE = 1
    INVALID_PARAM = 2
    OVERFLOW = 3
    NO_DATA = 4
    INVALID_HEADER = 5
    INCOMPLETE_PACKET = 6
    INVALID_LENGTH = 7
    INVALID_CHECKSUM = 8
    UNSET_PACKET = 9
    INVALID_TYPE = 10
    INVALID_SIZE = 11

class _DataInfo:
    def __init__(self):
        self.is_active = False
        self.ptr: Optional[List[Any]] = None
        self.type: Optional[TYPE] = None
        self.size = {"raw": 0, "encoded": 0}
        self.data = {"bits": 0, "encoded": 0}

class _DataSet:
    def __init__(self):
        self.is_editable = True
        self.id = 0
        self.info = [_DataInfo() for _ in range(MAX_DATA_NUM)]
        self.bit_length = 0     # 全データのエンコードビット長の合計
        self.payload_len = 0    # payload バイト長（bit_length から決まる）

class DataPacket:
    def __init__(self):
        self.id = 0
        # encoder 側では「フレーム全バイト数」、decoder 側では「header〜tail を含む長さ」を入れてよい
        self.length = 0
        self.data = bytearray(MAX_BYTE)

# ---------- Encoder (C++ dataEncoder 相当：1ポート専用) ----------

class DataCodecBaseEnc:
    def __init__(self, id_: int = 0):
        self._ds = _DataSet()
        self._ds.id = id_
        self._packet = DataPacket()
        self._packet.id = id_
        self._binary = [0]*MAX_BIT

    @staticmethod
    def _default_bits(tp: TYPE) -> int:
        if tp == TYPE.BOOL:  return 1
        if tp in (TYPE.UINT, TYPE.INT): return 32
        if tp == TYPE.FLOAT: return 32   # C++: FLOAT は常に32bit
        return 0

class DataEncoder(DataCodecBaseEnc):
    def __init__(self, id_: int = 0):
        super().__init__(id_)

    def append(self, ord_: int, ptr: List[Any], tp: TYPE, bits: Optional[int]=None) -> ERROR:
        if not (0 <= ord_ < MAX_DATA_NUM):
            return ERROR.INVALID_PARAM
        if not (isinstance(ptr, list) and len(ptr) == 1):
            return ERROR.INVALID_PARAM
        if not self._ds.is_editable:
            return ERROR.UNEDITABLE

        info = self._ds.info[ord_]
        if info.is_active:
            return ERROR.INVALID_PARAM

        info.is_active = True
        info.ptr = ptr
        info.type = tp

        # C++: FLOAT は常に32bit。それ以外は size or default
        if tp == TYPE.FLOAT:
            raw_bits = 32
        else:
            raw_bits = bits if (bits and bits > 0) else self._default_bits(tp)
        if raw_bits <= 0 or raw_bits > 64:
            return ERROR.INVALID_SIZE

        info.size["raw"] = raw_bits
        info.size["encoded"] = raw_bits
        info.data["bits"] = 0
        info.data["encoded"] = 0
        return ERROR.OK

    def set(self) -> ERROR:
        """bit ではなく payload バイト数で DATA_LENGTH_LIMIT をチェック"""
        total_bits = 0
        for i in range(MAX_DATA_NUM):
            if self._ds.info[i].is_active:
                total_bits += self._ds.info[i].size["encoded"]
        if total_bits == 0:
            return ERROR.NO_DATA

        payload_len = (total_bits + 6) // 7  # 7bit pack → バイト数

        if payload_len > DATA_LENGTH_LIMIT:
            return ERROR.OVERFLOW

        self._ds.bit_length = total_bits
        self._ds.payload_len = payload_len
        self._ds.is_editable = False
        return ERROR.OK

    def _gen_binary(self):
        idx = 0
        for i in range(MAX_DATA_NUM):
            inf = self._ds.info[i]
            if not inf.is_active:
                continue
            nbits = inf.size["encoded"]
            tp = inf.type
            if tp == TYPE.BOOL:
                u = 1 if bool(inf.ptr[0]) else 0
                width = 1
            elif tp == TYPE.UINT:
                u = int(inf.ptr[0]) & ((1 << nbits) - 1)
                width = nbits
            elif tp == TYPE.INT:
                u = int(inf.ptr[0]) & ((1 << nbits) - 1)
                width = nbits
            elif tp == TYPE.FLOAT:
                # float / double ともに float32 に丸めて送る（C++ヘッダ準拠）
                b = struct.pack('<f', float(inf.ptr[0]))
                u = int.from_bytes(b, 'little', signed=False)
                width = 32
            else:
                raise ValueError('invalid type')
            for j in range(width):
                bit = (u >> (width-1-j)) & 1
                self._binary[idx] = bit
                idx += 1

    def encode(self) -> ERROR:
        """C++ dataEncoder::encode() と同じフレーム構造で _packet を作る"""
        ds = self._ds
        if ds.is_editable or ds.bit_length == 0:
            return ERROR.UNSET_PACKET

        p = self._packet
        # 全クリア
        for i in range(MAX_BYTE):
            p.data[i] = 0

        # ビット列生成
        self._gen_binary()

        # header 10iiiiii
        p.data[0] = 0b10000000 | (ds.id & 0x3F)

        # payload (7bit pack)
        idx = 0  # payload index (data[idx+1])
        for i in range(ds.bit_length):
            bit = self._binary[i] & 1
            p.data[idx+1] |= (bit << (6 - (i % 7)))
            if (i % 7) == 6:
                idx += 1
        if (ds.bit_length % 7) != 0:
            idx += 1  # 最後の不完全バイト

        payload_len = idx  # payload バイト数

        # checksum（payloadのみの和）
        check = 0
        for i in range(1, payload_len+1):
            check += p.data[i]
        if check > 0x3FFF:
            return ERROR.OVERFLOW
        p.data[payload_len+1] = (check >> 7) & 0x7F
        p.data[payload_len+2] = check & 0x7F

        # tail 11llllll
        p.data[payload_len+3] = 0b11000000 | (payload_len % 64)

        # 総バイト数 = header(1) + payload + chk2 + tail(1)
        p.length = payload_len + 4
        return ERROR.OK

    def get_packet(self) -> DataPacket:
        """C++ の getPacket() と同様に DataPacket を返す"""
        return self._packet


# ---------- Decoder (C++ dataDecoder 相当：マルチポート) ----------

class DataDecoder:
    def __init__(self, id_: int = 0):
        # Single fixed packet ID decoder.
        self._data_set = _DataSet()
        self._data_set.id = id_
        self._data_set.is_editable = True

        self._packet = DataPacket()             # 直近のフレーム（header〜tail 含む）
        self._buffer = bytearray(BUFFER_SIZE)   # 受信バッファ
        self._buf_len = 0
        self._binary = [0]*MAX_BIT

    @staticmethod
    def _default_bits(tp: TYPE) -> int:
        if tp == TYPE.BOOL:  return 1
        if tp in (TYPE.UINT, TYPE.INT): return 32
        if tp == TYPE.FLOAT: return 32
        return 0

    def append(self, ord_: int, ptr: List[Any], tp: TYPE, bits: Optional[int]=None) -> ERROR:
        """Register a field for the fixed decoder ID."""
        if not (0 <= ord_ < MAX_DATA_NUM):
            return ERROR.INVALID_PARAM
        if not (isinstance(ptr, list) and len(ptr) == 1):
            return ERROR.INVALID_PARAM

        ds = self._data_set
        if not ds.is_editable:
            return ERROR.UNEDITABLE
        info = ds.info[ord_]
        if info.is_active:
            return ERROR.INVALID_PARAM

        info.is_active = True
        info.ptr = ptr
        info.type = tp

        if tp == TYPE.FLOAT:
            raw_bits = 32
        else:
            raw_bits = bits if (bits and bits > 0) else self._default_bits(tp)
        if raw_bits <= 0 or raw_bits > 64:
            return ERROR.INVALID_SIZE

        info.size["raw"] = raw_bits
        info.size["encoded"] = raw_bits
        info.data["bits"] = 0
        info.data["encoded"] = 0
        return ERROR.OK

    def set(self) -> ERROR:
        """Finalize the single decoder layout and validate payload size."""
        ds = self._data_set
        ds.bit_length = 0
        for i in range(MAX_DATA_NUM):
            if ds.info[i].is_active:
                ds.bit_length += ds.info[i].size["encoded"]

        if ds.bit_length == 0:
            ds.payload_len = 0
            ds.is_editable = False
            return ERROR.OK

        payload_len = (ds.bit_length + 6) // 7
        if payload_len > DATA_LENGTH_LIMIT:
            return ERROR.OVERFLOW

        ds.payload_len = payload_len
        ds.is_editable = False
        return ERROR.OK

    def get_current_id(self) -> int:
        """Return the last decoded packet ID."""
        return self._packet.id

    def append_to_buffer(self, b: int) -> ERROR:
        """Serial.read() 等で受けた1バイトを溜める"""
        if self._buf_len >= BUFFER_SIZE:
            # オーバーフロー時は半分だけ残して前半捨てるなどしても良いが，ここでは全クリア
            self._buf_len = 0
        self._buffer[self._buf_len] = b & 0xFF
        self._buf_len += 1
        return ERROR.OK

    def _shift_left(self, n: int):
        """バッファ先頭 n バイトを捨てる"""
        if n <= 0:
            return
        if n >= self._buf_len:
            self._buf_len = 0
            return
        remain = self._buf_len - n
        self._buffer[:remain] = self._buffer[n:self._buf_len]
        self._buf_len = remain

    def _extract_packet(self) -> ERROR:
        """バッファ先頭から1フレーム切り出して self._packet に格納"""
        if self._buf_len < 4:
            return ERROR.INCOMPLETE_PACKET

        # header 探索（10xxxxxx）
        if (self._buffer[0] & 0b11000000) != 0b10000000:
            idx = 1
            while idx < self._buf_len and (self._buffer[idx] & 0b11000000) != 0b10000000:
                idx += 1
            if idx == self._buf_len:
                # すべて捨てる
                self._buf_len = 0
                return ERROR.NO_DATA
            self._shift_left(idx)
            if self._buf_len < 4:
                return ERROR.INCOMPLETE_PACKET

        # tail 探索（11xxxxxx）
        length = 0
        while True:
            if 3 + length >= self._buf_len:
                return ERROR.INCOMPLETE_PACKET
            if (self._buffer[3+length] & 0b11000000) == 0b11000000:
                break
            length += 1
            if length > DATA_LENGTH_LIMIT:
                # 先頭1バイトだけ捨てて再同期
                self._shift_left(1)
                return ERROR.OVERFLOW

        # tail の l チェック
        tail = self._buffer[3+length]
        if (tail & 0x3F) != (length % 64):
            self._shift_left(1)
            return ERROR.INVALID_LENGTH

        # checksum (payload の和)
        check_calc = 0
        for i in range(length):
            check_calc += self._buffer[1+i]
        check_calc &= 0x3FFF
        c_hi = self._buffer[1+length] & 0x7F
        c_lo = self._buffer[2+length] & 0x7F
        check_recv = ((c_hi << 7) | c_lo) & 0x3FFF
        if check_calc != check_recv:
            self._shift_left(1)
            return ERROR.INVALID_CHECKSUM

        # フレーム確定：header〜tail をそのまま packet.data にコピー
        frame_len = length + 4
        self._packet.length = frame_len
        for i in range(frame_len):
            self._packet.data[i] = self._buffer[i]

        # id 抽出
        self._packet.id = self._buffer[0] & 0x3F

        # バッファからフレームを捨てる
        self._shift_left(frame_len)
        return ERROR.OK

    def _payload_bit_unpack(self):
        """payload 部を bit 列に戻す（_binary に格納）"""
        ds = self._data_set
        idx = 0
        for i in range(ds.bit_length):
            bindex = 1 + (i // 7)       # header の次から payload
            bitpos = 6 - (i % 7)
            bit = (self._packet.data[bindex] >> bitpos) & 1
            self._binary[idx] = bit
            idx += 1

    def _collect_encoded_words(self):
        """_binary から各データの encoded 値を復元"""
        ds = self._data_set
        idx = 0
        for i in range(MAX_DATA_NUM):
            inf = ds.info[i]
            if not inf.is_active:
                continue
            n = inf.size["encoded"]
            acc = 0
            for j in range(n):
                acc |= (self._binary[idx] << (n-1-j))
                idx += 1
            inf.data["encoded"] = acc

    def _get_bits_data(self) -> ERROR:
        """INT 用に符号拡張して bits を作る"""
        ds = self._data_set
        for i in range(MAX_DATA_NUM):
            inf = ds.info[i]
            if not inf.is_active:
                continue
            nraw = inf.size["raw"]
            enc = inf.data["encoded"]

            if inf.type == TYPE.INT:
                # 任意ビット幅の2の補数（1..64bit）
                if not (1 <= nraw <= 64):
                    return ERROR.INVALID_SIZE
                sign_bit = 1 << (nraw - 1)
                if nraw == 64:
                    mask = 0xFFFFFFFFFFFFFFFF
                else:
                    mask = (1 << nraw) - 1
                val = enc & mask
                if val & sign_bit:
                    full = val | (~mask)
                else:
                    full = val
                inf.data["bits"] = full
            else:
                inf.data["bits"] = enc
        return ERROR.OK

    def _restore_values(self) -> ERROR:
        """実際の変数に値を書き戻す"""
        ds = self._data_set
        for i in range(MAX_DATA_NUM):
            inf = ds.info[i]
            if not inf.is_active:
                continue
            nraw = inf.size["raw"]
            bits = inf.data["bits"]

            if inf.type == TYPE.BOOL:
                inf.ptr[0] = bool(bits & 1)

            elif inf.type == TYPE.UINT:
                if not (1 <= nraw <= 64):
                    return ERROR.INVALID_SIZE
                if nraw == 64:
                    mask = 0xFFFFFFFFFFFFFFFF
                else:
                    mask = (1 << nraw) - 1
                inf.ptr[0] = int(bits & mask)

            elif inf.type == TYPE.INT:
                if not (1 <= nraw <= 64):
                    return ERROR.INVALID_SIZE
                if nraw == 64:
                    mask = 0xFFFFFFFFFFFFFFFF
                else:
                    mask = (1 << nraw) - 1
                sign_bit = 1 << (nraw - 1)
                val = bits & mask
                if val & sign_bit:
                    full = val | (~mask)
                else:
                    full = val
                inf.ptr[0] = int(full)

            elif inf.type == TYPE.FLOAT:
                # 常に float32 から復元（double も float32 に丸めた値）
                temp = bits & 0xFFFFFFFF
                f = struct.unpack('<f', struct.pack('<I', temp))[0]
                if nraw in (32, 64):
                    inf.ptr[0] = float(f)
                else:
                    return ERROR.INVALID_SIZE

            else:
                return ERROR.INVALID_TYPE
        return ERROR.OK

    def decode(self) -> ERROR:
        """バッファから1フレーム分をデコードして，対応する id の変数群に書き込む"""
        err = self._extract_packet()
        if err != ERROR.OK:
            return err

        if self._packet.id != self._data_set.id:
            return ERROR.INVALID_PARAM

        ds = self._data_set
        if ds.is_editable or ds.bit_length == 0:
            return ERROR.UNSET_PACKET

        self._payload_bit_unpack()
        self._collect_encoded_words()
        err = self._get_bits_data()
        if err != ERROR.OK:
            return err
        err = self._restore_values()
        return err